storage: fix lease verification of lease request commands

petermattis opened the pull request:
Exalate commented:

**Describe the problem**

A [PR](https://github.com/cockroachdb/cockroach/issues/71102) was merged and released in CRDB v21.2.0 that added a check to the client cert authentication that ensured that the CN matches the username of the user we are authenticating as. It would be good to document this change in the release notes for 21.2.0 for others who may encounter this change.

**To Reproduce**

While less than stellar, we had a cron job that was using the node certificate bundle to authenticate as the root user. While this isn't optimal, upgrading our clusters to 21.2.0 suddenly broke our job. After some research, we found that the PR for issue 711202 changed the check, and switching to use the root user cert package instead resolved the issue.

**Expected behavior**
The change to validate the CN of the certificate bundle against the username should be documented for anyone who might encounter this in the future.

**Additional data / screenshots**
If the problem is SQL-related, include a copy of the SQL query and the schema
of the supporting tables.

If a node in your cluster encountered a fatal error, supply the contents of the
log directories (at minimum of the affected node(s), but preferably all nodes).

Note that log files can contain confidential information. Please continue
creating this issue, but contact support@cockroachlabs.com to submit the log
files in private.

If applicable, add screenshots to help explain your problem.

**Environment:**
 - CockroachDB version 21.1.9 upgraded to 21.2.0
 - Server OS: Linux
 - Client App: JDBC

**Additional context**
A cron job using the node certificate bundle to authenticate as root failed to authenticate suddenly after the upgrade to v21.2.0


***
If you all welcome outside PRs to documentation, simply point me in the right direction and I am happy to give it a shot :D
***


Jira Issue: DOC-1862

Jira issue: CRDB-11396

petermattis committed:
storage: fix lease verification of lease request commands

Fixes #12446

tamird commented:
There's a bunch of context in #12446 that should be present in the commit message here - we shouldn't have to go refer to that issue to understand this change. LGTM otherwise.

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/storage/replica.go, line 3280 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/12448#-KZ7shlm-wmDTua3EG9i:-KZ7shlm-wmDTua3EG9j:bpl70c5) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c55556ef14b323c8076f3f5f08ddad772f30746/pkg/storage/replica.go#L3280)):*
> ```Go
> 		if raftCmd.BatchRequest.IsSingleSkipLeaseCheckRequest() {
> 			// For lease commands, no additional verification is needed if the origin
> 			// lease isn't valid.
> ```

Could you provide an explanation?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12448#-:-KZ7v6X3mAawKNTHPsbh:b-6q1kzz)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
we discussed this offline; I don't think this is correct. Looking more.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/12448#-:-KZ7wM9BXoHRU2MDdUIz:bhyiyal)*
<!-- Sent from Reviewable.io -->

spencerkimball commented on pkg/storage/replica.go:
```diff
@@ -3275,6 +3275,13 @@ func (r *Replica) processRaftCommand(
+		if raftCmd.BatchRequest.IsSingleSkipLeaseCheckRequest() {
+			// For lease commands, no additional verification is needed if the origin
+			// lease isn't valid.
+			if !r.isLeaseValidLocked(raftCmd.OriginLease, ts) {
```
What prompted this change? It's definitely not correct however, as the results of `r.isLeaseValidLocked` can be different between different nodes, as node liveness information which is gossiped can easily be different. In other words, one replica might see an incremented epoch and decide there's no point in checking the lease and the other might not see it and go ahead and check the lease. Then you'd end up with a consistency error and perhaps worse, potentially two replicas which believe they have the lease.

petermattis replied:
The test failure in `TestRefreshPendingCommands` prompted it. Per @andreimatei's comment, this likely isn't correct, but something bad is happening here. The scenario is that this node (node 3) issued a `RequestLease` command when the previous expired lease was held by node 1. This `RequestLease` command is failing on the `Equivalent` check below because the current lease is held by node 3. Take a look at the log output attached to #12446.

petermattis committed:
storage: deflake TestRefreshPendingCommands

Detecting leadership change via changes to raft.Status.SoftState.Lead is
somewhat less robust than other mechanisms but is used because it can
often detect a new leader earlier.

spencerkimball commented on pkg/storage/replica.go:
```diff
@@ -3301,6 +3308,12 @@ func (r *Replica) processRaftCommand(
 			"command %s proposed from replica %+v: %s",
 			raftCmd.BatchRequest, raftCmd.OriginReplica, err,
 		)
+		// TODO(spencer,andrei): This error is confusing if r.mu.state.Lease is
```
We should probably return something different in that case. Or else update the code in `roachpb/errors.go` for `NotLeaseHolderError` to print something more sensible if the lease replica and replica are the same.

andreimatei replied:
I think the thing to do here is to return `LeaseRejectedError`, and I think we'll need to augment the code in `redirectOnOrAcquire` that interprets that error and currently converts it into a `notLeaseHolderError` to maybe recognize that the error might mean that in fact we the lease that we wanted might have been applied (in case the `LeaseRejectedError` is coming from a re-proposal).

petermattis commented:
Closing this PR as the approach isn't correct. @andreimatei is working on a correct fix.
